### PR TITLE
Adding UvA-DL notebook "Initialization and Optimization"

### PR DESCRIPTION
**Pull request by phlippe**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
      => Discussed on Slack channel
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?

This PR adds a new tutorial of the UvA Deep Learning course, this time on Initialization and Optimization. The tutorial is an adapted version of the original tutorial for the Deep Learning course at the University of Amsterdam ([link](https://uvadlc-notebooks.readthedocs.io/en/latest/tutorial_notebooks/tutorial4/Optimization_and_Initialization.html)).

The svg file is included in the notebook.

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py [hunk 1]
```diff
@@ -0,0 +1,1152 @@
...
+    return os.path.join(model_path, model_name + ".tar")
+
+
+def _get_result_file(model_path, model_name):
+    return os.path.join(model_path, model_name + "_results.json")
+
+
+def load_model(model_path, model_name, net=None):
+    config_file, model_file = _get_config_file(model_path,
+                                               model_name), _get_model_file(model_path, model_name)
```
Comment (Borda) [inline course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py:577]:
```suggestion
    config_file = _get_config_file(model_path, model_name)
    model_file = _get_model_file(model_path, model_name)
```

File: course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py [hunk 2]
```diff
@@ -0,0 +1,1152 @@
...
+    assert os.path.isfile(
+        config_file
+    ), f"Could not find the config file \"{config_file}\". Are you sure this is the correct path and you have your model config stored here?"
+    assert os.path.isfile(
+        model_file
+    ), f"Could not find the model file \"{model_file}\". Are you sure this is the correct path and you have your model stored here?"
```
Comment (Borda) [inline course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py:583]:
```suggestion
    assert os.path.isfile(
        config_file
    ), f'Could not find the config file "{config_file}". Are you sure this is the correct path and you have your model config stored here?'
    assert os.path.isfile(
        model_file
    ), f'Could not find the model file "{model_file}". Are you sure this is the correct path and you have your model stored here?'
```

File: course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py [hunk 3]
```diff
@@ -0,0 +1,1152 @@
...
+
+def _get_model_file(model_path, model_name):
+    return os.path.join(model_path, model_name + ".tar")
+
+
+def _get_result_file(model_path, model_name):
+    return os.path.join(model_path, model_name + "_results.json")
+
+
+def load_model(model_path, model_name, net=None):
```
Comment (Borda) [inline course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py:575]: seeing it in other notebooks, what do you think about having some common module for sharing?
**Pro**: do not repeat the same functions
**Con**: notebook wont be standalone any more

File: course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py [hunk 4]
```diff
@@ -0,0 +1,1152 @@
...
+        net = BaseNetwork(act_fn=act_fn, **config_dict)
+    net.load_state_dict(torch.load(model_file))
+    return net
+
+
+def save_model(model, model_path, model_name):
+    config_dict = model.config
+    os.makedirs(model_path, exist_ok=True)
+    config_file, model_file = _get_config_file(model_path,
+                                               model_name), _get_model_file(model_path, model_name)
```
Comment (Borda) [inline course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py:599]:
```suggestion
    config_file = _get_config_file(model_path, model_name)
    model_file = _get_model_file(model_path, model_name)
```

File: course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py [hunk 5]
```diff
@@ -0,0 +1,1152 @@
...
+            val_acc = test_model(net, val_loader)
+            val_scores.append(val_acc)
+            print(
+                f"[Epoch {epoch+1:2i}] Training accuracy: {train_acc*100.0:05.2f}%, Validation accuracy: {val_acc*100.0:05.2f}%"
+            )
+
+            if len(val_scores) == 1 or val_acc > val_scores[best_val_epoch]:
+                print("\t   (New best performance, saving model...)")
+                save_model(net, CHECKPOINT_PATH, model_name)
+                best_val_epoch = epoch
```
Comment (Borda) [inline course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py:670]: maybe cut this as a new function and call it inside to reduce complexity?

**Review by Borda — APPROVED** [review]

File: course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py [hunk 6]
```diff
@@ -0,0 +1,1152 @@
...
+import torch.utils.data as data
+# %matplotlib inline
+from IPython.display import set_matplotlib_formats
+from matplotlib import cm
+from torchvision import transforms
+from torchvision.datasets import FashionMNIST
+from tqdm.notebook import tqdm
+
+set_matplotlib_formats('svg', 'pdf')  # For export
+sns.set()
```
[the remark this answers is not in the document]
Comment (phlippe) [reply course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py]: Yes, agree, that would be great!

File: course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py [hunk 7]
```diff
@@ -0,0 +1,1152 @@
...
+# ### Constant variance
+#
+# From the experiment above, we have seen that a constant value is not working.
+# So instead, how about we initialize the parameters by randomly sampling from a distribution like a Gaussian?
+# The most intuitive way would be to choose one variance that is used for all layers in the network.
+# Let's implement it below, and visualize the activation distribution across layers.
+
+
+# %%
+def var_init(model, std=0.01):
```
[the remark this answers is not in the document]
Comment (phlippe) [reply course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py:386]: The mean is assumed to be zero, I made it more explicit in the initialization.

File: course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py [hunk 8]
```diff
@@ -0,0 +1,1152 @@
...
+# Let's implement it below and check whether this holds:
+
+
+# %%
+def equal_var_init(model):
+    for name, param in model.named_parameters():
+        if name.endswith(".bias"):
+            param.data.fill_(0)
+        else:
+            param.data.normal_(std=1.0 / math.sqrt(param.shape[1]))
```
[the remark this answers is not in the document]
Comment (phlippe) [reply course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py:451]: The above functions apply a certain initialization to _all_ parameters. Here, we apply the different initialization for single parameter tensors, which does not fit in the previous function definitions.

File: course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py (hunk 3 above)
Comment (phlippe) [reply course_UvA-DL/initialization-and-optimization/Initialization_and_Optimization.py:582]: I was also thinking about it, but we only have two notebooks that currently share some part of their code. I would prefer to keep the notebooks stand-alone and independent of each other, but I am also ok with making a shared module.

**Pull request merged**